server/commands/write_authzmodel.go

```diff
@@ -163,3 +180,83 @@ func isUsersetRewriteValid(allRelations map[string]struct{}, relationsOnType map
 
 	return nil
 }
+
+func areRelationalTypesValid(tds []*openfgapb.TypeDefinition) error {
+	typeToRelations := map[string]map[string]struct{}{}
```

Comment:
There should be a helper method for this in the `typesystem` package.

I think we should have something like

```
package typesystem

type Typesystem struct {
    model *openfgapb.AuthorizationModel
}

func New(model *openfga.AuthorizationModel) *Typesystem {
    ....
}

func (t *Typesystem) GetTypeRelationsMap() map[string]map[string]Relation {
    // logic here
}
```

Reply:
Thanks! As discussed, let's refactor in a follow up PR.

Reply:
I left a comment here:
https://github.com/openfga/openfga/pull/223#discussion_r974494567

I don't think this work should get put off, because it's part of the review of this code. I feel we should make the composition of this code more readable before merging, because it makes the code easier to read and maintain and also lays the foundation for further developments to make use of more reusable software pieces related to type definitions.

The last remaining bit that I feel should be done is just a slight refactor like I recommended below. I think once we make it a little more readable through better composition, then we're good to merge 👍